src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.MethodParametersInfo.cs

```diff
@@ -0,0 +1,305 @@
+            /// void Callee(int a, int b, params bool[] c)
+            /// {
+            ///     DoSomething(a, b, c);
+            /// }
+            /// After:
+            /// void Caller(int i, int j, bool[] k)
+            /// {
+            ///     DoSomething(i, j, k);
+            /// }
+            /// void Callee(int a, int b, params bool[] c)
+            /// {
+            ///     DoSomething(a, b, c);
+            /// }
+            /// </summary>
+            public ImmutableDictionary<IParameterSymbol, string> ParametersWithIdentifierArgument { get; }
+
+            /// <summary>
+            /// Parameters map to variable declaration argument's name.
+            /// This is only used for C# to support the 'out' variable declaration. For VB it should always be empty.
+            /// Before:
+            /// void Caller()
+            /// {
+            ///     Callee(out var x);
+            /// }
+            /// void Callee(out int i) => i = 100;
+            ///
+            /// After:
+            /// void Caller()
+            /// {
+            ///     int x = 100;
+            /// }
+            /// void Callee(out int i) => i = 100;
+            /// </summary>
+            public ImmutableArray<(IParameterSymbol parameterSymbol, string)> ParametersWithVariableDeclarationArgument { get; }
+
+            /// <summary>
+            /// Operations that represent Parameter has argument but the argument is not identifier or literal.
+            /// For these parameters they are considered to be put into a declaration statement after inlining.
+            /// Note: params array could maps to multiple/zero arguments.
+            /// Example:
+            /// Before:
+            /// void Caller(bool x)
+            /// {
+            ///     Callee(Foo(), x ? Foo() : Bar())
+            /// }
+            /// void Callee(int a, int b)
+            /// {
+            ///     DoSomething(a, b);
+            /// }
+            /// After:
+            /// void Caller(bool x)
+            /// {
+            ///     int a = Foo();
+            ///     int b = x ? Foo() : Bar();
+            ///     DoSomething(a, b);
+            /// }
+            /// void Callee(int a, int b)
+            /// {
+            ///     DoSomething(a, b);
+            /// }
+            /// </summary>
+            public ImmutableArray<IArgumentOperation> OperationsToGenerateFreshVariablesFor { get; }
+
+            /// <summary>
+            /// Parameters has no argument input and have default value. They will be replaced by their default value.
+            /// </summary>
+            public ImmutableArray<IParameterSymbol> ParametersWithDefaultValue { get; }
+
+            /// <summary>
+            /// Parameters map to literal expression argument. They will be replaced by the literal value.
+            /// </summary>
+            public ImmutableDictionary<IParameterSymbol, TExpressionSyntax> ParametersWithLiteralArgument { get; }
```

Comment:
this seems practically the same as ParametersWithIdentifierArgument.  Should we merge into one map ParametersWithLiteralOrIdentifierArgument?

Reply:
Emm, I think it actually it could also be used for the default value, identifierArgument & the special case we discussed.
Updated the implementation.